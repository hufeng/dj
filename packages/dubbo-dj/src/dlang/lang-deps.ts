export type TFromModule = string
export interface IDepValue {
  is3rdLib: boolean
  defaultModule: string
  noneDefaultModules: Set<string>
}
export interface IDepAdd {
  fromModule: string
  importModule: string
  isDefaultModule?: boolean
  is3rdLib?: boolean
}

export class DubboDep {
  private readonly depMap: Map<TFromModule, IDepValue>

  constructor() {
    this.depMap = new Map<TFromModule, IDepValue>()
  }

  add({
    fromModule,
    importModule,
    isDefaultModule = false,
    is3rdLib = false,
  }: IDepAdd) {
    if (this.depMap.has(fromModule)) {
      const moduleValue = this.depMap.get(fromModule) || {
        defaultModule: '',
        noneDefaultModules: new Set(),
      }
      if (isDefaultModule) {
        moduleValue.defaultModule = importModule
      } else {
        moduleValue.noneDefaultModules.add(importModule)
      }
    } else {
      if (isDefaultModule) {
        this.depMap.set(fromModule, {
          is3rdLib,
          defaultModule: importModule,
          noneDefaultModules: new Set<string>(),
        })
      } else {
        this.depMap.set(fromModule, {
          is3rdLib,
          defaultModule: '',
          noneDefaultModules: new Set([importModule]),
        })
      }
    }
  }

  getDepMap() {
    return this.depMap
  }
}
