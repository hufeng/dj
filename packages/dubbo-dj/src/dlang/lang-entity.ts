import Lang from './lang'
import { universalType } from './type'
import { IDubboEntityField, IType, TDataType } from '../types'

// ~~~~~~~~~~~~~~~~~~ model~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEntity extends Lang {
  fields: Array<IDubboEntityField>

  constructor(fullName: string, comment?: string) {
    super(fullName, comment)
    this.fields = []
    this.deps.add({
      fromModule: 'js-to-java',
      isDefaultModule: true,
      importModule: 'java',
      is3rdModule: true,
    })
  }

  field(name: string, type: IType, comment?: string) {
    this.fields.push({
      comment: this.wrapComment(comment),
      name,
      type,
    })
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.comment,
      fullName: this.fullName,
      shortName: this.shortName,
      infName: this.infName,
      fields: this.fields,
      imports: this.imports(this.deps),
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class EntityBuilder {
  private readonly entity: DubboEntity

  constructor(fullName: string, comment?: string) {
    this.entity = new DubboEntity(fullName, comment)
  }

  field(name: string, type: TDataType, comment?: string) {
    const t = universalType(this.entity.deps, type)
    this.entity.field(name, t, comment)
    return this
  }

  ok() {
    return this.entity
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~
export function entity(fullName: string, comment?: string) {
  return new EntityBuilder(fullName, comment)
}
