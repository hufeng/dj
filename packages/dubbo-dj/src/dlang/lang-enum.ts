import Lang from './lang'
import { IDubboEnumFiled } from '../types'

// ~~~~~~~~~~~~~~~~~~ model~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEnum extends Lang {
  fields: Array<IDubboEnumFiled> = []

  constructor(fullName: string, comment?: string) {
    super(fullName, comment)
  }

  field(name: string, val: number | string, comment?: string) {
    this.fields.push({
      name,
      val,
      comment: this.wrapComment(comment),
    })
    return this
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.comment,
      fullName: this.fullName,
      fields: this.fields,
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEnumDSL {
  private readonly enumeration: DubboEnum

  constructor(fullName: string, comment: string = '') {
    this.enumeration = new DubboEnum(fullName, comment)
  }

  /**
   * 枚举属性字段
   * @param name 属性名称
   * @param val
   * @param comment
   */
  field(name: string, val: number | string, comment?: string) {
    this.enumeration.field(name, val, comment)
    return this
  }

  /**
   * 完成Enum DSL的定义
   */
  ok() {
    return this.enumeration
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~
export function enumeration(cls: string, comment?: string) {
  return new DubboEnumDSL(cls, comment)
}
