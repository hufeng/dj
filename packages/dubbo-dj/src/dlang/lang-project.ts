import { IProject } from './type'

export class DubboProject {
  private readonly meta: IProject = { name: '', version: '', description: '' }

  name(name: string) {
    this.meta.name = name
    return this
  }

  version(version: string) {
    this.meta.version = version
    return this
  }

  description(description: string) {
    this.meta.description = description
    return this
  }

  ok() {
    return this.meta
  }
}

export const project = new DubboProject()
