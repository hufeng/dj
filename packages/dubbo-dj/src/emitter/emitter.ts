import debug from 'debug'
import fsx from 'fs-extra'
import { fmt } from '../common'

const FILE_EXT = {
  ts: 'ts',
  go: 'go',
  java: 'java',
}
const log = debug(`dj:emitter：say ~`)

export default abstract class Emitter {
  protected readonly genFilePath: string
  private readonly fullName: string

  protected constructor(
    fullName: string,
    lang: 'ts' | 'go' | 'java' = 'ts',
    baseDir: string = './dubbo'
  ) {
    this.fullName = fullName
    const fileExt = FILE_EXT[lang]
    const pkgDir = this.dotPath(fullName)
    this.genFilePath = `${baseDir}/${lang}/${pkgDir}.${fileExt}`
  }

  dotPath(cls: string) {
    return cls.replace(/\./g, '/')
  }

  // write to => ./dubbo/ts/**/*.ts
  async writeCode(): Promise<void> {
    try {
      await fsx.ensureFile(this.genFilePath)
      await fsx.writeFile(this.genFilePath, this.prettyCode)
      log('create %s => %s was successfully', this.fullName, this.genFilePath)
    } catch (err) {
      log('write file error %s', err)
      console.error(err)
    }
  }

  get prettyCode(): string {
    return fmt(this.code)
  }

  /**
   * 子类只需要实现该方法即可
   */
  abstract get code(): string
}
