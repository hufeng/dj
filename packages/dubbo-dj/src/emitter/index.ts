import DubboEnumEmitter from './emitter-enum'
import DubboEntityEmitter from './emitter-entity'
import DubboServiceProviderEmitter from './emitter-provider-service'
import DubboAbstractServiceEmitter from './emitter-abstract-service'
import DubboConsumerService from './emitter-consumer-service'

export {
  DubboEnumEmitter,
  DubboEntityEmitter,
  DubboConsumerService,
  DubboServiceProviderEmitter,
  DubboAbstractServiceEmitter,
}
