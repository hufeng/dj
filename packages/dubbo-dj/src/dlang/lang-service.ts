import Lang from './lang'
import { universalType } from './type'
import { IArg, IFuncArg, IType, TDataType, TFuncOption } from '../types'
import { DubboDep } from './lang-deps'

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~ model ~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboService extends Lang {
  funcs: Array<Func>

  constructor(fullName: string, comment: string = '') {
    super(fullName, comment)
    this.funcs = []
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.wrapComment(this.comment),
      fullName: this.fullName,
      shortName: this.shortName,
      infName: this.infName,
      funcs: this.funcs,
      imports: this.imports(this.deps),
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class ServiceBuilder {
  private readonly service: DubboService

  constructor(fullName: string, comment: string = '') {
    this.service = new DubboService(fullName, comment)
  }

  func(...f: Array<TFuncOption>) {
    const func = new Func()
    for (let opt of f) {
      opt(func, this.service.deps)
    }
    this.service.funcs.push(func)
    return this
  }

  get ok() {
    return this.service
  }
}

export function service(name: string, comment?: string) {
  return new ServiceBuilder(name, comment)
}

// ~~~~~~~~~ Func && FuncBuilder ~~~~~~~~~~~~~~~~~~
export class Func {
  public comment: string = ''
  public name: string = ''
  public args: Array<IFuncArg> = []
  public ret: IType = null as any
}

// ~~~~~~~~~~~~~~~ functional option type ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export const f = {
  name: (name: string, comment: string = '') => (func: Func) => {
    func.name = name
    func.comment = comment
  },
  ret: (type: TDataType) => (func: Func, deps: DubboDep) => {
    func.ret = universalType(deps, type)
  },
  args: (...args: Array<IArg>) => (func: Func, deps: DubboDep) => {
    func.args = args.map((v) => {
      return {
        ...v,
        type: universalType(deps, v.type),
      }
    })
  },
  arg: (name: string, type: TDataType, comment?: string) => {
    return {
      name,
      type,
      comment,
    }
  },
}
