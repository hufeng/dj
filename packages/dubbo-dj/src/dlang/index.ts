import { entity } from './lang-entity'
import { service, f } from './lang-service'
import { enumeration } from './lang-enum'
import {
  Integer,
  String,
  Boolean,
  List,
  boolean,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  char,
  Char,
  chars,
  Character,
  Currency,
} from './type'

export const t = {
  Boolean,
  boolean,
  Integer,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  String,
  char,
  Char,
  chars,
  Character,
  List,
  Set,
  Currency,
}

export { entity, service, enumeration, f }
