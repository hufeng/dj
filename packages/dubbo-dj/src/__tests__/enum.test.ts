import { enumeration } from '../dlang'
import { DubboEnumEmitter } from '../emitter'

describe('enumeration test suite', () => {
  it('test color number enum', () => {
    const color = enumeration('org.apache.dubbo.e.Color')
      .field('Red', 0, '红色')
      .field('Green', 1, '绿色')
      .field('Blue', 2, '蓝色')
      .ok()

    expect(color.fullName).toEqual('org.apache.dubbo.e.Color')
    expect(color.comment).toEqual('')
    expect(color.fields.length).toEqual(3)
    expect(color.fields.map((c) => ({ name: c.name, val: c.val }))).toEqual([
      { name: 'Red', val: 0 },
      { name: 'Green', val: 1 },
      { name: 'Blue', val: 2 },
    ])
    expect(color).toMatchSnapshot()
  })

  it('test string enum', () => {
    const fileType = enumeration('org.apache.dubbo.e.FileType')
      .field('PDF', 'PDF', 'PDF')
      .field('MP3', 'MP3', 'mp3')
      .ok()
    expect(fileType.fullName).toEqual('org.apache.dubbo.e.FileType')
    expect(fileType.comment).toEqual('')
    expect(fileType.fields.length).toEqual(2)
    expect(fileType.fields.map((c) => ({ name: c.name, val: c.val }))).toEqual([
      { name: 'PDF', val: 'PDF' },
      { name: 'MP3', val: 'MP3' },
    ])
    expect(fileType).toMatchSnapshot()
    expect(new DubboEnumEmitter(fileType, 'ts').prettyCode).toMatchSnapshot()
  })
})
