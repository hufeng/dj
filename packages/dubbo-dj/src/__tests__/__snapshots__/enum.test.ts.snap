// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enumeration test suite test color number enum 1`] = `
DubboEnum {
  "comment": "",
  "deps": DubboDep {
    "depMap": Map {},
  },
  "fields": Array [
    Object {
      "comment": "
    /**
    * Á∫¢Ëâ≤
    */",
      "name": "Red",
      "val": 0,
    },
    Object {
      "comment": "
    /**
    * ÁªøËâ≤
    */",
      "name": "Green",
      "val": 1,
    },
    Object {
      "comment": "
    /**
    * ËìùËâ≤
    */",
      "name": "Blue",
      "val": 2,
    },
  ],
  "fullName": "org.apache.dubbo.e.Color",
  "infName": "IColor",
  "shortName": "Color",
}
`;

exports[`enumeration test suite test string enum 1`] = `
DubboEnum {
  "comment": "",
  "deps": DubboDep {
    "depMap": Map {},
  },
  "fields": Array [
    Object {
      "comment": "
    /**
    * PDF
    */",
      "name": "PDF",
      "val": "PDF",
    },
    Object {
      "comment": "
    /**
    * mp3
    */",
      "name": "MP3",
      "val": "MP3",
    },
  ],
  "fullName": "org.apache.dubbo.e.FileType",
  "infName": "IFileType",
  "shortName": "FileType",
}
`;

exports[`enumeration test suite test string enum 2`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ üíó machine coding üíó ~~~
 */

export enum FileType {
  /**
   * PDF
   */
  PDF = 'PDF',

  /**
   * mp3
   */
  MP3 = 'MP3',
}
"
`;
